kaggle_environments/envs/lux_ai_2022/luxai2022/factory.py

```diff
@@ -0,0 +1,158 @@
+                if pos_lichen >= MIN_LICHEN_TO_SPREAD:
+                    seen.add(tuple(check_pos))
+                    frontier.append(check_pos)
+            elif adj_strain != strain_id:
+                    # adjacent tile is not empty and is not a strain this factory owns.
+                    can_grow = False
+                    seen.add(tuple(check_pos))
+            else:
+                # adjacent tile has our own strain, we can grow here too
+                seen.add(tuple(check_pos))
+                frontier.append(check_pos)
+
+        if can_grow:
+            grow_lichen_positions.add((pos[0], pos[1]))
+    return grow_lichen_positions
+
+class Factory:
+    def __init__(self, team: Team, unit_id: str, num_id: int) -> None:
+        self.team_id = team.team_id
+        self.team = team
+        self.unit_id = unit_id
+        self.pos = Position(np.zeros(2, dtype=int))
+        self.power = 0
+        self.cargo = UnitCargo()
+        self.num_id = num_id
+        self.action_queue = [] # TODO can we queue actions or are factories outside of max control limit
+        self.grow_lichen_positions = None
+
+    @property
+    def pos_slice(self):
+        return slice(self.pos.y - 1, self.pos.y + 2), slice(self.pos.x - 1, self.pos.x + 2)
+
+    def refine_step(self, config: EnvConfig):
+        consumed_ice = min(self.cargo.ice, config.FACTORY_PROCESSING_RATE_WATER)
+        consumed_ore = min(self.cargo.ore, config.FACTORY_PROCESSING_RATE_METAL)
+
+        self.cargo.ice -= consumed_ice
+        self.cargo.ore -= consumed_ore
+
+        # TODO - are we rounding or doing floats or anything?
+        self.cargo.water += int(consumed_ice / config.ICE_WATER_RATIO)
+        self.cargo.metal += int(consumed_ore / config.ORE_METAL_RATIO)
+
+    def cache_water_info(self, board: Board, env_cfg: EnvConfig):
+        # TODO this can easily be a fairly slow function, can we make it much faster?
+        # Caches information about which tiles lichen can grow on for this factory
+
+        # perform a BFS from the factory position and look for non rubble, non factory tiles.
+        # find the current frontier from 4 starting positions x marked below
+        """
+             x
+           _ _ _
+          |     |
+        x |     | x
+          |_ _ _|
+             x
+
+        """
+        forbidden = (board.rubble > 0) | (board.factory_occupancy_map != -1)
+        init_arr = np.stack([self.pos.pos + np.array([0, -2]), self.pos.pos + np.array([2, 0]), self.pos.pos + np.array([0, 2]), self.pos.pos + np.array([-2, 0])])
+        self.grow_lichen_positions = compute_water_info(init_arr, env_cfg.MIN_LICHEN_TO_SPREAD, board.lichen, board.lichen_strains, self.num_id, forbidden)
+    def water_cost(self, config: EnvConfig):
+        return int(np.ceil(len(self.grow_lichen_positions) / config.LICHEN_WATERING_COST_FACTOR) + 1)
+
+    ### Add and sub resource functions copied over from unit.py code, can we consolidate them somewhere?
```

Comment:
I think the unit code will be different enough given the constraints on a units cargo